packages/request-idle-callback-polyfill/src/__tests__/RequestIdleCallback-test.js

```diff
@@ -0,0 +1,143 @@
+  const animationFrameStart = highResolutionTimer++;
+  animationFrameCallbacks.length = 0;
+  callbacksToRun.forEach(cb => cb(animationFrameStart));
+}
+
+function mockLongRunningCode() {
+  highResolutionTimer += 100;
+}
+
+describe('RequestIdleCallback', () => {
+  beforeAll(() => {
+    // When error supression is enabled, jest is not reporting expect failures
+    // inside of idle callbacks.
+    Error.prototype.suppressReactErrorLogging = false;
+    previousRAF = window.requestAnimationFrame;
+    previousPostMessage = window.postMessage;
+    window.postMessage = postMessage;
+    window.performance = {now: performanceNow};
+    window.requestAnimationFrame = requestAnimationFrame;
+  });
+
+  afterAll(() => {
+    window.requestAnimationFrame = previousRAF;
+    window.postMessage = previousPostMessage;
+    previousPostMessage = null;
+    previousRAF = null;
+    delete window.performance;
+  });
+
+  beforeEach(() => {
+    animationFrameCallbacks = [];
+    highResolutionTimer = 0xf000;
+    jest.resetModules();
+    requestIdleCallback = require('request-idle-callback-polyfill')
+      .requestIdleCallback;
+    cancelIdleCallback = require('request-idle-callback-polyfill')
+      .cancelIdleCallback;
+  });
+
+  describe('requestIdleCallback', () => {
+    it('returns a number', () => {
+      const callback = jest.fn();
+      expect(typeof requestIdleCallback(callback)).toBe('number');
+    });
+    it('executes callbacks asynchronously', () => {
+      const callback = jest.fn();
+      requestIdleCallback(callback);
+      expect(callback).not.toBeCalled();
+      mockRunNextFrame();
+      expect(callback).toBeCalled();
+    });
+    it('cancels callbacks', () => {
+      const callback = jest.fn();
+      const handle = requestIdleCallback(callback);
+      cancelIdleCallback(handle);
+      mockRunNextFrame();
+      expect(callback).not.toBeCalled();
+    });
+
+    it('passes a deadline to the callback', () => {
+      const callback = jest.fn(deadline => {
+        expect(deadline.didTimeout).toBe(false);
+        expect(deadline.timeRemaining()).toBeGreaterThan(0);
+        mockLongRunningCode();
+        expect(deadline.timeRemaining()).toBe(0);
+      });
+      requestIdleCallback(callback);
+      mockRunNextFrame();
+      expect(callback).toBeCalled();
+    });
+
+    it('stops executing callbacks if the deadline expires', () => {
+      const ops = [];
+      requestIdleCallback(() => ops.push('first'));
+      requestIdleCallback(() => {
+        ops.push('second');
+        mockLongRunningCode();
+      });
+      requestIdleCallback(() => ops.push('third'));
+      expect(ops).toEqual([]);
+      mockRunNextFrame();
+      expect(ops).toEqual(['first', 'second']);
+      mockRunNextFrame();
+      expect(ops).toEqual(['first', 'second', 'third']);
+    });
+
+    it('executes callbacks that timeout', () => {
+      const callback = jest.fn(deadline => {
+        expect(deadline.didTimeout).toBe(true);
```

Comment:
Our typical strategy is to never put assertions inside of callbacks. It's too easy to neglect to call the callback. Instead, we push into a log and make assertions on it in the main function. See one of our async tests for an example: https://github.com/facebook/react/blob/48ffbf06beb272d2ab9484b1084063757b6a64f1/packages/react-reconciler/src/__tests__/ReactIncremental-test.internal.js#L75-L140

Then you can get rid of `suppressErrorLogging = false`.

Reply:
It does assert that the callback is called, but this looks much better than hacking `suppressErrorLogging` 👌

Reply:
Another benefit of a log is the test will fail if there are multiple invocations.